registries/adapters/galaxy_adapter.go

```diff
@@ -0,0 +1,259 @@
+)
+
+const (
+	defaultRunner     = "ansibleplaybookbundle/apb-base:latest"
+	galaxyName        = "galaxy"
+	galaxySearchURL   = "https://galaxy-qa.ansible.com/api/v1/content/?content_type__name=apb"
+	galaxyNSSearchURL = "https://galaxy-qa.ansible.com/api/v1/content/?content_type__name=apb&namespace__name=%v"
+	galaxyRoleURL     = "https://galaxy-qa.ansible.com/api/v1/content/%v/"
+	galaxyAPIURL      = "https://galaxy-qa.ansible.com/api/v1"
+)
+
+// GalaxyAdapter - Galaxy Adapter
+type GalaxyAdapter struct {
+	Config Configuration
+}
+
+// GalaxyRole - Role from Ansible Galaxy.
+type GalaxyRole struct {
+	Name    string            `json:"name"`
+	RoleID  int               `json:"id"`
+	Summary GalaxyRoleSummary `json:"summary_fields"`
+}
+
+// GalaxyRoleResponse - Role Response from Ansible Galaxy.
+type GalaxyRoleResponse struct {
+	Name     string             `json:"name"`
+	Metadata GalaxyRoleMetadata `json:"metadata"`
+	Summary  GalaxyRoleSummary  `json:"summary_fields"`
+}
+
+// GalaxyRoleMetadata - Role Metadata obtained from Role Response.
+type GalaxyRoleMetadata struct {
+	Spec bundle.Spec `json:"apb_metadata"`
+}
+
+// GalaxyRoleSummary - Role Summary obtained from Role Response.
+type GalaxyRoleSummary struct {
+	Namespace GalaxyRoleNamespace `json:"namespace"`
+}
+
+// GalaxyRoleNamespace - Role Namespace obtained from Role Response Summary.
+type GalaxyRoleNamespace struct {
+	Name string `json:"name"`
+}
+
+// GalaxySearchResponse - Search response for Galaxy.
+type GalaxySearchResponse struct {
+	Count   int           `json:"count"`
+	Results []*GalaxyRole `json:"results"`
+	Next    string        `json:"next"`
+}
+
+// RegistryName - Retrieve the registry name
+func (r GalaxyAdapter) RegistryName() string {
+	return galaxyName
+}
+
+// GetImageNames - retrieve the images
+func (r GalaxyAdapter) GetImageNames() ([]string, error) {
+	log.Debug("GalaxyAdapter::GetImages")
+	log.Debug("BundleSpecLabel: %s", BundleSpecLabel)
+	log.Debug("Loading role list with tag: [apb]")
+
+	channel := make(chan string)
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	defer cancelFunc()
+
+	var imageResp *GalaxySearchResponse
+	var err error
+	// Intial call to getNextImages this will fan out to retrieve all the values.
+	if len(r.Config.Org) == 0 {
+		imageResp, err = r.getNextImages(ctx, galaxySearchURL, channel, cancelFunc)
+	} else {
+		imageResp, err = r.getNextImages(ctx, fmt.Sprintf(galaxyNSSearchURL, r.Config.Org), channel, cancelFunc)
+	}
+	// if there was an issue with the first call, return the error
+	if err != nil {
+		return nil, err
+	}
+
+	// If no results in the fist call then close the channel as nothing will get loaded.
+	if len(imageResp.Results) == 0 {
+		log.Info("canceled retrieval as no items in org")
+		close(channel)
+	}
+	var apbData []string
+	counter := 1
+	for imageData := range channel {
+		apbData = append(apbData, imageData)
+		if counter < imageResp.Count {
+			counter++
+		} else {
+			close(channel)
```

Comment:
This feels weird to do in a for loop. But I guess it's the only way to exit the loop. Just spidey senses, nothing concrete wrong with it.

Reply:
I'm a little concerned as well? I think of the channel as a shared resource that should have one owner. It's not uncommon to have an outer fn make a channel and then pass it to an inner fn (effectively passing ownership over to the inner fn), and then for looping over the channel range until the inner fn has closed the channel. It feels a little weird to me here that both the outer, and inner fn seem to claim ownership over when this channel should be closed, thus ending the loop.